src/software/new_geom/cubic_bezier_spline2d.cpp

```diff
@@ -0,0 +1,167 @@
+    for (size_t i = 0; i < num_knots; i++)
+    {
+        knot_vector.emplace_back(i * 1.0 / static_cast<double>(num_knots - 1));
+    }
+    return knot_vector;
+}
+
+const Point CubicBezierSpline2d::getStartPoint() const
+{
+    // We are guaranteed to at least have two control points, the start and end points
+    return control_points.front();
+}
+
+const Point CubicBezierSpline2d::getEndPoint() const
+{
+    // We are guaranteed to at least have two control points, the start and end points
+    return control_points.back();
+}
+
+const std::vector<SplineSegment2d> CubicBezierSpline2d::getSplineSegments() const
+{
+    std::vector<SplineSegment2d> segments;
+    const double num_knots = getNumKnots();
+    for (size_t i = 0; i < num_knots - 1; i++)
+    {
+        const BezierCurve2d curve = getSegmentAtIndex(i);
+        segments.emplace_back(createSplineSegment2d(0, 1, curve.getPolynomial()));
+    }
+    return segments;
+}
+
+BezierCurve2d CubicBezierSpline2d::getSegmentAtIndex(size_t index) const
+{
+    return BezierCurve2d({control_points[index * 3], control_points[index * 3 + 1],
+                          control_points[index * 3 + 2], control_points[index * 3 + 3]});
+}
+
+std::vector<Point> CubicBezierSpline2d::computeControlPoints(
+    const Point& start_point, const Vector& start_vector, const Point& end_point,
+    const Vector& end_vector, std::vector<Point> intermediate_knots)
+{
+    // The math here is based off of: https://www.ibiblio.org/e-notes/Splines/b-int.html
+    // And the variable names have been maintained (where reasonably possible) to match
+    // those found at this aforementioned link.
+
+    std::vector<Point> all_knots;
+    all_knots.emplace_back(start_point);
+    all_knots.insert(all_knots.end(), intermediate_knots.begin(),
+                     intermediate_knots.end());
+    all_knots.emplace_back(end_point);
+
+    const size_t n = all_knots.size();
+
+    std::vector<Vector> A(n - 2, Vector(0, 0));
+    std::vector<double> B(n - 2, 0);
+
+    if (n - 2 > 0)
+    {
+        B[0] = -0.25;
+        A[0] = (all_knots[2] - all_knots[0] - start_vector) / 4;
+    }
+    for (size_t i = 1; i < A.size(); i++)
+    {
+        A[i] = (all_knots[i + 2] - all_knots[i] - A[i - 1]) / (4 + B[i - 1]);
+        B[i] = -1.0 / (4.0 + B[i - 1]);
+    }
+
+    std::vector<Vector> d(n, Vector(0, 0));
+    d[0]     = start_vector;
+    d[n - 1] = -end_vector;
+
+    for (size_t i = n - 2; i > 0; i--)
+    {
+        d[i] = A[i - 1] + B[i - 1] * d[i + 1];
+    }
+
+    std::vector<Point> control_points;
+    for (size_t i = 0; i < all_knots.size(); i++)
+    {
+        if (i > 0)
+        {
+            control_points.emplace_back(all_knots[i] - d[i]);
+        }
+        control_points.emplace_back(all_knots[i]);
+        if (i < n - 1)
+        {
+            control_points.emplace_back(all_knots[i] + d[i]);
+        }
+    }
+
+    assert((control_points.size() - 1) % 3 == 0);
```

Comment:
Just double checking we want an assert here rather than another type of failure? (I understand if we hit this something has gone horrifically mathematically wrong)

Reply:
Yeah, this should never be wrong, just makes debugging easier.